catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswEndpoint.java

```diff
@@ -536,46 +554,103 @@ public TransactionResponseType transaction(CswTransactionRequest request) throws
     response.setVersion(CswConstants.VERSION_2_0_2);
 
     int numInserted = 0;
+    final Subject subject = SecurityUtils.getSubject();
     for (InsertAction insertAction : request.getInsertActions()) {
-      insertAction = transformInsertAction(insertAction);
-      CreateRequest createRequest = new CreateRequestImpl(insertAction.getRecords());
-      try {
-        CreateResponse createResponse = framework.create(createRequest);
-        if (request.isVerbose()) {
-          response.getInsertResult().add(getInsertResultFromResponse(createResponse));
-        }
+      final InsertAction transformInsertAction = transformInsertAction(insertAction);
+
+      List<Metacard> metacards = transformInsertAction.getRecords();
+      CompletionService<CreateResponse> completionService =
+          new ExecutorCompletionService<>(queryExecutor);
+
+      for (Metacard record : metacards) {
+        CreateRequest createRequest = new CreateRequestImpl(record);
+        Callable<CreateResponse> callable =
+            () -> {
+              try {
+                return framework.create(createRequest);
+              } catch (IngestException | SourceUnavailableException e) {
+                LOGGER.debug("Unable to insert record(s)", e);
+                throw new CswException(
+                    "Unable to insert record(s).",
+                    CswConstants.TRANSACTION_FAILED,
+                    transformInsertAction.getHandle());
+              }
+            };
+        Callable<CreateResponse> createCallable = subject.associateWith(callable);
+        completionService.submit(createCallable);
+      }
 
-        numInserted += createResponse.getCreatedMetacards().size();
-      } catch (IngestException | SourceUnavailableException e) {
-        LOGGER.debug("Unable to insert record(s)", e);
-        throw new CswException(
-            "Unable to insert record(s).",
-            CswConstants.TRANSACTION_FAILED,
-            insertAction.getHandle());
+      for (int i = 0; i < metacards.size(); i++) {
+        try {
+          Future<CreateResponse> completedFuture = completionService.take();
+
+          try {
+            CreateResponse futureResponse = completedFuture.get();
+            numInserted += futureResponse.getCreatedMetacards().size();
+            if (request.isVerbose()) {
+              response.getInsertResult().add(getInsertResultFromResponse(futureResponse));
+            }
+          } catch (ExecutionException | CancellationException e) {
+            LOGGER.debug("Error ingesting Metacard", e);
+            throw new CswException(
+                "Unable to insert record(s).",
+                CswConstants.TRANSACTION_FAILED,
+                insertAction.getHandle());
+          }
+        } catch (InterruptedException e) {
+          LOGGER.debug("Metacard ingest interrupted", e);
+          Thread.currentThread().interrupt();
```

Comment:
So correct me if i'm just misinterpreting this, but 
- if we are interrupted (and an InterruptedException is thrown) it seems like we are currently 
    1) Setting the interrupt flag on the thread (👍), 
    2) Going to continue running to the next iteration of the for loop but possibly  ignoring the future that was grabbed (and theoretically it could get interrupted during the `take` in which case an iteration of the for loop would be advanced and an item would be left on the queue -- but since we're serially adding all items to the completion service and then starting to take them off I don't think we can get into that particular situation) 

Meaning even though we are setting the interrupt flag again we are continuing with execution so we aren't really stopping even though we were asked to. Am I misinterpreting/misunderstanding this? (Will calling interrupt() somehow result in the thread aborting?). It would seem if we want to respect being interrupted we could stop processing, and return (the flow of control upward, until someone who is handling interrupts can deal with the now set flag. Not sure how this is dealt with in the calling code (i believe CXF is what we're using for jax-rs).

Reply:
Added a break after setting the current thread interrupted.